core-services/src/main/java/org/zalando/nakadi/service/publishing/NakadiRecordMapper.java

```diff
@@ -44,35 +41,26 @@ public List<NakadiRecord> fromBytesBatch(final byte[] batch) throws IOException
 
             final Schema metadataSchema = avroSchema.getEventTypeSchema(
                     AvroSchema.METADATA_KEY, Byte.toString(metadataVersion));
-            final GenericDatumReader datumReader = new GenericDatumReader(metadataSchema);
-            final GenericRecord genericRecord = (GenericRecord) datumReader.read(null,
-                    DecoderFactory.get().directBinaryDecoder(
-                            new ByteArrayInputStream(metadata), null));
             records.add(new NakadiRecord()
-                    .setMetadata(new GenericRecordMetadata(genericRecord, metadataVersion))
+                    .setMetadata(new NakadiAvroMetadata(metadataVersion, metadataSchema, metadata))
                     .setData(wholeRecord));
         }
 
         return records;
     }
 
-    public NakadiRecord fromAvroGenericRecord(final byte metadataVersion,
-                                              final GenericRecord metadata,
+    public NakadiRecord fromAvroGenericRecord(final NakadiAvroMetadata metadata,
                                               final GenericRecord event) throws IOException {
         final byte[] data = EnvelopeHolder.produceBytes(
-                metadataVersion,
-                (outputStream -> {
-                    final GenericDatumWriter eventWriter = new GenericDatumWriter(metadata.getSchema());
-                    eventWriter.write(metadata, EncoderFactory.get()
-                            .directBinaryEncoder(outputStream, null));
-                }),
+                metadata.getMetadataVersion(),
+                metadata,
```

Comment:
Oh.  Now I'm not sure I like the implicit way that `metadata` implements the writer interface...